samples/96b_argonkey: divided test into sensor and microphone (#11188)
zephyrproject-rtos/zephyr: samples/boards/96b_argonkey/microphone/README.rst

.. _ArgonKey_Microphone:

ArgonKey Board Microphone
#########################

Overview
********
This sample provides an example of how to acquire audio through
the on-board MP34DT05 microphone. The microphone generates a PDM
stream which is acquired through I2S. The PDM stream is then
converted to PCM using the OpenPDMFilter library, which is available
in source code in this sample.

Requirements
************

This sample requires the ArgonKey board plus a USB to TTL 1V8 serial
cable to get the output audio stream. The board can be powered
in either one of the following two ways:

- mezzanine mode, plugging the ArgonKey to HiKey board thru its 96Board
  low-speed connector
- standalone mode, supplying 5V directly on P1 connector

References
**********

- :ref:`96b_argonkey`

Building and Running
********************

>  .. zephyr-app-commands::

Reviewer: Sorry, missed this in the previous review.  Directives, such as ``.. zephyr-app-command`` and ``.. code-block`` should be indented at the same level as the paragraphs around them.  For the directive here on line 33, there should be no leading spaces.

Author: ok

.. _ArgonKey_Microphone:

ArgonKey Board Microphone
#########################

Overview
********
This sample provides an example of how to acquire audio through
the on-board MP34DT05 microphone. The microphone generates a PDM
stream which is acquired through I2S. The PDM stream is then
converted to PCM using the OpenPDMFilter library, which is available
in source code in this sample.

Requirements
************

> This sample just requires the ArgonKey board. The board can be fed

Author: I'll put 'powered' instead of 'fed' globally.
I'll add also the cable requirement.

in either one of the following two ways:

- mezzanine mode, plugging the ArgonKey to HiKey board thru its 96Board
  low-speed connector

> - standalone mode, feeding directly 5V on P1 connector

Author: Yes.

> - standalone mode, feeding directly 5V on P1 connector

Author: Well, actually it doesn't sound good to me to repeat the word 'power'.
I can probably say:

"The board can be powered [.......] PROVIDING 5V directly on P1 connector."

What do you think?

> - standalone mode, feeding directly 5V on P1 connector

Author: Maybe SUPPLYING ?

References
**********

- :ref:`96b_argonkey`

Building and Running
********************

 .. zephyr-app-commands::
    :zephyr-app: samples/boards/96b_argonkey/microphone
    :host-os: unix
    :board: 96b_argonkey
    :goals: run
    :compact:

Sample Output
=============

The example acquires 3s of audio and prints out the PCM stream on COM port.
A USB to TTL 1V8 serial cable may be attached to the low speed connector on
the back of the board on P3.5 (TX) and P3.7 (RX).

> As soon as the acquisition starts the green led glows. At the end of the

Reviewer: ```suggestion
As soon as the acquisition starts the green LED glows. At the end of the
```

> acquisition both the green and red led goes on and the PCM stream is sent

Reviewer: ```suggestion
acquisition both the green and red LEDs go on and the PCM stream is sent
```

on COM port. When the output is completed the green led goes off and the red
led stays on. The process can be re-iterated only by pressing the RST button.

- acquisition starts:     GRN on  - RED off
- acquisition ends:       GRN on  - RED on
- output ends:            GRN off - RED on

> The characteristic of the PCM audio are hardcoded in the example:

Reviewer: ```suggestion
The characteristics of the PCM audio are hardcoded in the example:
```

> - 32KHz of sample rate

Reviewer: ```suggestion
- 32KHz sample rate
```

- 16 bits per sample
- 1 channel (mono), as only 1 microphone is available

> 3 seconds of acquisition @32Kz of sampling rate corresponds to 96000 sample

Reviewer: ```suggestion
Three seconds of acquisition at a 32KHz sampling rate yields 96,000 16-bit samples.
```

(x 16bits each).

Since the MP34DT05 microphone is PDM, we need to apply an oversampling factor,
which is 64 in this example, resulting in a 2MHz bit clock. During the PCM

> conversion the OpenPDMFilter library is filtering out and decimating the data

Reviewer: ```suggestion
conversion, the OpenPDMFilter library filters out and decimates the data
```

by the same 64 factor.

At the end of the acquisition the PCM data will be printed on the terminal

> emulator in either binary or ascii format. The output is controlled by

Reviewer: ```suggestion
emulator in either binary or ASCII format. The output is controlled by
```

> PCM_OUTPUT_IN_ASCII definition, which is off by default.

Reviewer: mention where this define is declared (samples/boards/96b_argonkey/microphone/src/main.c)

Binary PCM Output
-----------------

> The ttyUSB0 port must be configured in raw mode in order to avoid to have

Reviewer: ```suggestion
The ttyUSB0 port must be configured in raw mode to avoid having
```

characters 'cooked' out.

>  .. code-block:: console

Reviewer: This code block (and contents) are indented by one extra space.  (The ``.. code-block`` directive should start in column 1.

    stty -F /dev/ttyUSB0 115200 raw
    cat /dev/ttyUSB0 > /tmp/sound.raw

> NOTE!

Reviewer: Change this to:

    ..note:: In case the character 0x0a is interpreted as NL and an 0x0d (CR) is added,
       you may need to remove it::

          dos2unix -f /tmp/sound.raw

In case the character 0x0a is interpreted as NL and 0x0d (CR) is added
then it may be necessary to remove it:

 .. code-block:: console

    dos2unix -f /tmp/sound.raw

ASCII PCM Output
----------------

It is also possible to recompile and to have PCM output in ASCII, which needs

> to be converted in binary later on. The output format is the following:

Reviewer: ```suggestion
to be converted to binary later on. The output format is the following:
```

.. code-block:: console

    ArgonKey test!!
    -- start
    0xfbe0,
    0xfbf0,
    0xfc0c,
    0xfc24,
    0xfc3c,
    0xfc4c,
    0xfc68,
    0xfc48,

    [...]

    0xfb98,
    0xfb98,
    0xfbb8,
    0xfbac,
    0xfbc4,
    0xfbe8,
    0xfbf4,
    -- end

Play PCM Audio
--------------

> At this point in either case we ended up with a binary PCM file, say sound.raw.

Reviewer: Change this to:

    Now that we have a binary PCM file (say sound.raw), you can use,
    for example, the audicity open source editor/player to load and play it.

The audacity open source audio editor/player can be used to load and play the
PCM audio file.

> Just use the 'Import->Raw Data' item from menu and load the sound.raw as

Reviewer: ```suggestion
Use the 'Import->Raw Data' menu to load the sound.raw file as
```

- 16KHz sample rate
- 16 bits per sample
- 1 channel (mono), as only 1 microphone is available

Five seconds of acquisition at a 16KHz sampling rate yields 80,000 16-bit samples.
The microphone PDM requested clock should lead the MP34DT05 driver to select an
oversampling/decimation factor equal to 128, resulting in a 2.048MHz bit clock.

See configuration in file :file:`samples/boards/96b_argonkey/microphone/src/main.c`:

.. code-block:: c

   #define AUDIO_FREQ          16000
   #define CHAN_SIZE           16
   #define PCM_BLK_SIZE_MS     ((AUDIO_FREQ/1000) * sizeof(s16_t))

   #define NUM_MS              5000

   K_MEM_SLAB_DEFINE(rx_mem_slab, PCM_BLK_SIZE_MS, NUM_MS, 1);

   struct pcm_stream_cfg mic_streams = {
        .pcm_rate = AUDIO_FREQ,
        .pcm_width = CHAN_SIZE,
        .block_size = PCM_BLK_SIZE_MS,
        .mem_slab = &rx_mem_slab,
   };

   struct dmic_cfg cfg = {
        .io = {
            /* requesting a pdm freq around 2MHz */
            .min_pdm_clk_freq = 1800000,
            .max_pdm_clk_freq = 2500000,
        },
        .streams = &mic_streams,
        .channel = {
            .req_num_chan = 1,
        },

>    };

Reviewer 2: This is quite a lot of code for the README and will likely get out of sync with the actual c file. I think you should remove it.

.. note:: It is possible to change the AUDIO_FREQ to 32000 acquiring only 2500 ms. In this
   case the oversampling/decimation factor will be 64.

At the end of the acquisition the PCM data will be printed on the terminal
emulator in either binary or ASCII format. The output is controlled by
following macro, off by default, in :file:`samples/boards/96b_argonkey/microphone/src/main.c`:

* :c:macro:`PCM_OUTPUT_IN_ASCII`

Binary PCM Output
-----------------

The ttyUSB0 port must be configured in raw mode to avoid having

> characters 'cooked' out.

Reviewer 2: 'cooked'? Is this a well-known term?

Author: Yes. Basically TTY drivers has two modes 'raw' vs 'cooked'. In this sample we need raw otherwise some characters may be processed by the driver itself and not sent out, corrupting the audio stream (you will hear a lot of noise).
